NIFI-3446 FlowConfigurationDAO Changes

patricker opened the pull request:
Allows for Flow Configuration (flow.xml.gz) to be pluggable. There should be no "functional" changes as a result of this code change.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

pvillard31 commented on nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java:
```diff
@@ -48,6 +48,7 @@
 
     // core properties
     public static final String PROPERTIES_FILE_PATH = "nifi.properties.file.path";
+    public static final String FLOW_CONFIGURATION_IMPLEMENTATION = "nifi.flow.configuration.implementation";
```
This needs to be added in nifi.properties file in nifi-resources.

joewitt commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardFlowService.java:
```diff
@@ -225,33 +208,17 @@ private StandardFlowService(
-        } finally {
-            writeLock.unlock();
-        }
+        dao.save(controller);
```
this no longer appears to have write lock protection

joewitt commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardFlowService.java:
```diff
@@ -225,33 +208,17 @@ private StandardFlowService(
-        } finally {
-            writeLock.unlock();
-        }
+        dao.save(controller, outStream);
```
this no longer has write lock protection

joewitt commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardFlowService.java:
```diff
@@ -225,33 +208,17 @@ private StandardFlowService(
-        } finally {
-            writeLock.unlock();
-        }
+        dao.overwriteFlow(is);
```
this no longer has write lock protection

joewitt commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardFlowService.java:
```diff
@@ -284,7 +251,7 @@ public boolean isRunning() {
     @Override
     public void start() throws LifeCycleStartException {
-        writeLock.lock();
+       dao.writeProtectFlow();
```
this is leaking responsibilities.  The 'dao' should either be said to be thread safe or not.  The previous approach was not thread safe and thus we had locking around it as was in the previous implementation.  Here we have some cases guarded and some not but we're actually invoking a lock within the non-thread safe class now.  This approach is harder to reason over in my opinion.  We should either make the class thread safe and the caller does not need to do anything special or it should be not thread safe and the calling class should protect it (as it was).  Though this method shows we need to have some of that logic external so it makes sense to keep the locking as it was (controlled by the caller with the callee/dao being not thread safe by design)

patricker replied:
Thanks @joewitt, I'll move responsibility back over to the caller, along with your other feedback, and push an update.

joewitt commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/persistence/FlowConfigurationDAO.java:
```diff
@@ -25,11 +25,41 @@
+    /**
+     * Locks the Flow to prevent cross-thread Writeing
+     */
+    void writeProtectFlow();
```
i mentioned it in a previous comment but the need to make locking calls available by callers into this class indicates something a bit awkward happening.  See previous comments.

joewitt commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/persistence/FlowConfigurationDAO.java:
```diff
@@ -25,11 +25,41 @@
+    /**
+     * Locks the Flow to prevent cross-thread Writeing
+     */
+    void readProtectFlow();
```
copy/paste error on the read/unread protection javadocs.

patricker commented:
@joewitt @pvillard31 I made the changes and squashed the commits since about 50% of my changes were basically rolled back to the original code.

One issue... it doesn't work anymore. I'm looking for a little guidance on what I might have missed (of course since there is no functional change maybe it never worked and I was testing with an old build?). The error makes it sound like it's a standard, 'can't find your class' error. Since the code all lives in the same JAR I added a resource file to register the interface and class (it's in the commit), which made no difference.
Here is the stack trace I'm getting:

> Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flowService': FactoryBean threw exception on object creation; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flowController': FactoryBean threw exception on object creation; nested exception is java.lang.RuntimeException: Unable to create Flow Configuration DAO
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.doGetObjectFromFactoryBean(FactoryBeanRegistrySupport.java:175) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.getObjectFromFactoryBean(FactoryBeanRegistrySupport.java:103) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getObjectForBeanInstance(AbstractBeanFactory.java:1585) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:317) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1060) ~[spring-context-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.apache.nifi.web.contextlistener.ApplicationStartupContextListener.contextInitialized(ApplicationStartupContextListener.java:53) ~[na:na]
	... 28 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flowController': FactoryBean threw exception on object creation; nested exception is java.lang.RuntimeException: Unable to create Flow Configuration DAO
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.doGetObjectFromFactoryBean(FactoryBeanRegistrySupport.java:175) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.getObjectFromFactoryBean(FactoryBeanRegistrySupport.java:103) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getObjectForBeanInstance(AbstractBeanFactory.java:1585) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:317) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1060) ~[spring-context-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.apache.nifi.spring.StandardFlowServiceFactoryBean.getObject(StandardFlowServiceFactoryBean.java:48) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.doGetObjectFromFactoryBean(FactoryBeanRegistrySupport.java:168) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	... 34 common frames omitted
Caused by: java.lang.RuntimeException: Unable to create Flow Configuration DAO
	at org.apache.nifi.controller.FlowController.<init>(FlowController.java:480) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.FlowController.createStandaloneInstance(FlowController.java:378) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.spring.FlowControllerFactoryBean.getObject(FlowControllerFactoryBean.java:74) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.doGetObjectFromFactoryBean(FactoryBeanRegistrySupport.java:168) ~[spring-beans-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	... 41 common frames omitted
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: org.apache.nifi.persistence.StandardXMLFlowConfigurationDAO
	at org.apache.nifi.controller.FlowController.createFlowConfigurationDAO(FlowController.java:851) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.FlowController.<init>(FlowController.java:478) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	... 44 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.apache.nifi.persistence.StandardXMLFlowConfigurationDAO
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381) ~[na:1.8.0_91]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_91]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331) ~[na:1.8.0_91]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_91]
	at java.lang.Class.forName0(Native Method) ~[na:1.8.0_91]
	at java.lang.Class.forName(Class.java:264) ~[na:1.8.0_91]
	at org.apache.nifi.nar.NarThreadContextClassLoader.createInstance(NarThreadContextClassLoader.java:194) ~[nifi-nar-utils-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.FlowController.createFlowConfigurationDAO(FlowController.java:845) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	... 45 common frames omitted

patricker commented:
I'm going to just close it as unworkable right now. If someone wants to salvage this later feel free to use my code or start over.

joewitt commented:
thanks peter.  i think this will be much easier and better aligned with the intent as the flow registry work progresses.
